### PR TITLE
Enable leader election in csi components

### PR DESCRIPTION
**Pull request by bcho**
The following csi components provides builtin leader election support:

- csi-provisioner
- csi-attacher
- csi-snapshotter
- csi-resizer

To improve availability of controller service, we can enable leader
election in these components and allow the deployment to replica to 3.

for #177

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Release note**:
```

```

### CODE REVIEW

**Comment by andyzhangx** [comment]
so if we change controller component replica to 3, then there will be leader election process?

**Comment by bcho** [comment]
> 
> 
> so if we change controller component replica to 3, then there will be leader election process?

Yep, I had tested in local with `replica=2`, here are some sample logs:

Container that won the election:

```
I1219 08:35:20.049140       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:25.065304       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:30.103713       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:35.120147       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:40.139180       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:45.154194       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:50.168808       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:35:55.182444       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:36:00.196013       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
```

Container which is not holding the lease:

```
I1219 08:36:34.261370       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
I1219 08:36:34.261392       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:36:41.729194       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
I1219 08:36:41.729216       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
I1219 08:36:46.915125       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
I1219 08:36:46.915282       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
```

**Comment by andyzhangx** [comment]
> > so if we change controller component replica to 3, then there will be leader election process?
> 
> Yep, I had tested in local with `replica=2`, here are some sample logs:
> 
> Container that won the election:
> 
> ```
> I1219 08:35:20.049140       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:25.065304       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:30.103713       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:35.120147       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:40.139180       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:45.154194       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:50.168808       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:35:55.182444       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:36:00.196013       1 leaderelection.go:258] successfully renewed lease kube-system/external-attacher-leader-disk-csi-azure-com
> ```
> 
> Container which is not holding the lease:
> 
> ```
> I1219 08:36:34.261370       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
> I1219 08:36:34.261392       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:36:41.729194       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
> I1219 08:36:41.729216       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
> I1219 08:36:46.915125       1 leaderelection.go:326] lock is held by csi-azuredisk-controller-56875c4f68-d7txb and has not yet expired
> I1219 08:36:46.915282       1 leaderelection.go:222] failed to acquire lease kube-system/external-attacher-leader-disk-csi-azure-com
> ```

great, what num do you suggest 2 or 3? could you change the replic num in both charts and azuredisk-controller.yaml in this PR? Thanks.

**Comment by bcho** [comment]
> could you change the replic num in both charts and azuredisk-controller.yaml in this PR

Updated, PTAL

> what num do you suggest 2 or 3

I don't have a clear answer and I think this may depends on the number of the master nodes? (So I made the replicas count changeable in the chart.)

File: charts/latest/azuredisk-csi-driver/templates/csi-azuredisk-controller.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ metadata:
   namespace: {{ .Release.Namespace }}
 {{ include "azuredisk.labels" . | indent 2 }}
 spec:
-  replicas: 1
+  replicas: {{ .Values.controller.replicas }}
```
Comment (andyzhangx) [inline charts/latest/azuredisk-csi-driver/templates/csi-azuredisk-controller.yaml]: what's the default value now? this csi driver would run e2e test on this chart config, `3` is ok for me.

File: charts/latest/azuredisk-csi-driver/values.yaml [hunk 2]
```diff
@@ -37,3 +37,6 @@ serviceAccount:
 
 rbac:
   create: true
+
+controller:
+  replicas: 3
```
Comment (andyzhangx) [inline charts/latest/azuredisk-csi-driver/values.yaml]: I see, it's here, let's wait for the test result

**Comment by bcho** [comment]
>  W1219 09:38:55.963] merge: 031bfff78310f1ba4b1446454aaffc0782196a76 - not something we can merge 

Looks like it failed because my force push...

/retest

**Comment by andyzhangx** [comment]
since we have 3 controller pods now, the original script(https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/test/utils/azuredisk_log.sh) won't work any more? Could you fix that in this PR? Thanks.

```
SS2019/12/19 10:29:03 ===================azuredisk log===================
I1219 10:29:04.756] error: expected 'logs [-f] [-p] (POD | TYPE/NAME) [-c CONTAINER]'.
I1219 10:29:04.756] POD or TYPE/NAME is a required argument for the logs command
I1219 10:29:04.756] See 'kubectl logs -h' for help and examples
I1219 10:29:04.760] 
I1219 10:29:04.760] ------------------------------
I1219 10:29:04.760] Failure [0.917 seconds]
I1219 10:29:04.760] [AfterSuite] AfterSuite 
I1219 10:29:04.761] /go/src/sigs.k8s.io/azuredisk-csi-driver/test/e2e/suite_test.go:112
I1219 10:29:04.761] 
I1219 10:29:04.761]   Unexpected error:
I1219 10:29:04.761]       <*exec.ExitError | 0xc0009d5040>: {
I1219 10:29:04.761]           ProcessState: {
I1219 10:29:04.761]               pid: 60559,
I1219 10:29:04.761]               status: 256,
I1219 10:29:04.761]               rusage: {
I1219 10:29:04.762]                   Utime: {Sec: 0, Usec: 208444},
I1219 10:29:04.762]                   Stime: {Sec: 0, Usec: 127268},
I1219 10:29:04.762]                   Maxrss: 48328,
I1219 10:29:04.762]                   Ixrss: 0,
I1219 10:29:04.762]                   Idrss: 0,
I1219 10:29:04.762]                   Isrss: 0,
I1219 10:29:04.762]                   Minflt: 7496,
I1219 10:29:04.763]                   Majflt: 278,
I1219 10:29:04.763]                   Nswap: 0,
I1219 10:29:04.763]                   Inblock: 65624,
I1219 10:29:04.763]                   Oublock: 920,
I1219 10:29:04.763]                   Msgsnd: 0,
I1219 10:29:04.763]                   Msgrcv: 0,
I1219 10:29:04.763]                   Nsignals: 0,
I1219 10:29:04.763]                   Nvcsw: 2496,
I1219 10:29:04.764]                   Nivcsw: 809,
I1219 10:29:04.764]               },
I1219 10:29:04.764]           },
I1219 10:29:04.764]           Stderr: nil,
I1219 10:29:04.764]       }
I1219 10:29:04.764]       exit status 1
I1219 10:29:04.764]   occurred
I1219 10:29:04.764] 
I1219 10:29:04.764]   /go/src/sigs.k8s.io/azuredisk-csi-driver/test/e2e/suite_test.go:131
```

**Comment by bcho** [comment]
No problem I’ll update them later

**Comment by bcho** [comment]
> 
> 
> /lgtm
> /approve
> @bcho could you also port such code change to azurefile & blobfuse csi driver? Thanks.

👌 I will port these changes in next week,

**Pull request merged**